pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,744 @@
+	}
+
+	if cfg.CPU != "" {
+		baseargs = append(baseargs, "-smp", cfg.CPU)
+	} else {
+		baseargs = append(baseargs, "-smp", fmt.Sprintf("%d", runtime.NumCPU()))
+	}
+
+	// use kvm on linux, and Hypervisor.framework on macOS
+	if runtime.GOOS == "linux" {
+		baseargs = append(baseargs, "-enable-kvm")
+	} else if runtime.GOOS == "darwin" {
+		baseargs = append(baseargs, "-accel", "hvf")
+	}
+
+	baseargs = append(baseargs, "-cpu", "host")
+	baseargs = append(baseargs, "-daemonize")
+	// ensure we disable unneeded devices, this is less needed if we use microvm machines
+	// but still useful otherwise
+	baseargs = append(baseargs, "-display", "none")
+	baseargs = append(baseargs, "-no-reboot")
+	baseargs = append(baseargs, "-no-user-config")
+	baseargs = append(baseargs, "-nodefaults")
+	baseargs = append(baseargs, "-parallel", "none")
+	baseargs = append(baseargs, "-serial", "none")
+	baseargs = append(baseargs, "-vga", "none")
+	// use -netdev + -device instead of -nic, as this is better supported by microvm machine type
+	baseargs = append(baseargs, "-netdev", "user,id=id1,hostfwd=tcp::"+cfg.SSHPort+"-:22")
+	baseargs = append(baseargs, "-device", "virtio-net-pci,netdev=id1")
+	// add random generator via pci, improve ssh startup time
+	baseargs = append(baseargs, "-device", "virtio-rng-pci,rng=rng0", "-object", "rng-random,filename=/dev/urandom,id=rng0")
+	// panic=-1 ensures that if the init fails, we immediately exit the machine
+	// Add default SSH keys to the VM
+	sshkey := base64.StdEncoding.EncodeToString(pubKey)
+	baseargs = append(baseargs, "-append", "quiet nomodeset panic=-1 sshkey="+sshkey)
+	// we will *not* mount workspace using qemu, this will use 9pfs which is network-based, and will
+	// kill all performances (lots of small files)
+	// instead we will copy back the finished workspace artifacts when done.
+	// this dramatically improves compile time, making them comparable to bwrap or docker runners.
+	baseargs = append(baseargs, "-fsdev", "local,security_model=mapped,id=fsdev100,path="+cfg.WorkspaceDir)
+	baseargs = append(baseargs, "-device", "virtio-9p-pci,id=fs100,fsdev=fsdev100,mount_tag=defaultshare")
+
+	// if no size is specified, let's go for a default
+	if cfg.Disk == "" {
+		clog.FromContext(ctx).Infof("qemu: no disk space specified, using default: %s", defaultDiskSize)
+		cfg.Disk = defaultDiskSize
+	}
+
+	// if we want a disk, just add it, the init will mount it to the build home automatically
+	diskFile, err := generateDiskFile(ctx, cfg.Disk)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: could not generate additional disks: %v", err)
+		return err
+	}
+
+	// append raw disk, init will take care of formatting it if present.
+	baseargs = append(baseargs, "-object", "iothread,id=io1")
+	baseargs = append(baseargs, "-device", "virtio-blk-pci,drive=disk0,iothread=io1")
+	baseargs = append(baseargs, "-drive", "if=none,id=disk0,cache=none,format=raw,aio=threads,werror=report,rerror=report,file="+diskFile)
```

Comment:
per qemu-system-x86 [man page](https://manpages.ubuntu.com/manpages/noble/man1/qemu-system.1.html) :

> The  most explicit way to describe disks is to use a combination of -device to specify the
> hardware device and -blockdev to describe the backend. The device defines what  the  guest
> sees and the backend describes how QEMU handles the data. It is the only guaranteed stable
> interface for describing block devices and as such is recommended for management tools and
> scripting.
>
> The  -drive option combines the device and backend into a single command line option which
> is a more human friendly. There is however no interface stability guarantee although  some
> older board models still need updating to work with the modern blockdev forms.

So it would be better here to use a `-device` and `-blockdev`.  Unfortunately I don't have an example of that.

Reply:
Will test

`	baseargs = append(baseargs, "-blockdev", "file,node-name=disk0,locking=off,cache.direct=off,discard=ignore,aio=threads,filename="+diskFile)`

seems like it should cover the current command

Reply:
If this is non-trivial, then we can leave that to future.

Reply:
> Sadly -blockdev is really poorly documented :/

Yeah, my source of documentation has always been either source code or "what does libvirt do here".

I've just been bit by the -drive option before, having it not work in between versions.
its fine as TODO.